module/os/linux/zfs/zfs_dir.c

```diff
@@ -726,11 +722,17 @@ zfs_rmnode(znode_t *zp)
 		 * which point we'll call zfs_unlinked_drain() to process it).
 		 */
 		dmu_tx_abort(tx);
-		zfs_znode_dmu_fini(zp);
 		goto out;
 	}
 
 	if (xzp) {
+		/*
+		 * XXX what's the point of this assertion?
```

Comment:
Original Illumos commit for this is https://github.com/illumos/illumos-gate/commit/0a586cea3ceec7e5e50e7e54c745082a7a333ac2

Who would know? (@ahrens , @behlendorf ?)
Should I just remove this XXX?

Reply:
If anyone ever added a condition to or changed the "goto out:" in the preceding block, the assert would still document that the dmu_tx_assign needs to have succeeded for the "if (xzp)" block to work properly, so I'm ok with it.

Reply:
In that case the error should probably be a `const int tx_assign_error`. Because, right now, it's used a scratch variable for different error values.  Will factor that out into a separate patch.

Reply:
> Will factor that out into a separate patch.

Actually, I don't think it's worth it: we're uncertain about why the assertion is there. And, since the value lifetimes of the different values assigned to `error` in this function are non-overlapping, I think having multiple `const int OP_error` variables will just lead to `optimized out` values in GDB :P